juju/juju
# [JUJU-4369] AWS Ipv6 Support

With this PR we are enabling IPv6 for AWS in both dual and single stack. This PR also introduces new fixes to our instance selection logic to make instance type determination deterministic.

- We are now going to prefer starting instances in dual stack subnets
- We will talk to the controller from the client on ipv6 addresses
- We will make relevant ICMP and Juju firewall rules to allow ipv6 traffic.
- Juju will now only pick the highest generation general purpose AWS instance types now. We will no longer pick machines at will. This new approach will be deterministic and reliable while allowing us to use machines that support ipv6.

## Checklist


## QA steps

With this PR we are trying to assert 2 things. The first is that Juju will bootstrap and work in both single stack and dual stack environments. The second is we are looking to make sure that we have deterministic machine type selection.

Get this code base and compile everything. I have prepared some testing VPC's

### First Test Single Stack
1. Bootstrap Juju to the single stack VPC. This VPC has NAT64 enabled so the instance can reach the deb packages.
`juju bootstrap  --config vpc-id=vpc-042e10f8faf56df93  aws/ap-southeast-2`
2. You want to confirm the following:
- That only an ipv6 address is reported during bootstrap and used to make a connection to the controller.
- That a latest gen m series machine is used. This should be m6i.large for intel and m7g.medium for arm64
- Investigate the security groups on the machines to confirm that icmpv6 rules are created.
3. Add a model and a machine. Confirm the Juju agent is talking to the controller. This should all be via ipv6

I have left off deploying work loads as I am not sure we have any charms that will take single stack very well at the moment.

### Second Test Dual Stack
1. Bootstrap Juju to the single stack VPC. This VPC has NAT64 enabled so the instance can reach the deb packages.
`juju bootstrap  --config vpc-id=vpc-0dac80a34d244ef57  aws/ap-southeast-2 dualstack`
2. You want to confirm the following:
- That ipv4 and ipv6 addresses are returned during the bootstrap and that one of the addresses succeeds.
- That a latest gen m series machine is used. This should be m6i.large for intel and m7g.medium for arm64
- Investigate the security groups on the machines to confirm that icmpv6 rules are created.
3. Add a model and a machine. Confirm the Juju agent is talking to the controller. This should all be via itv

## Documentation changes

N/A

## Bug reference

https://bugs.launchpad.net/juju/+bug/1694988 (ipv6)
https://bugs.launchpad.net/juju/+bug/2029133 (deterministic machine types)
